Add dependency on parent type when indexing fields

gmorenz opened the pull request:
Fixes #570.

This takes the approach of just importing the modules that we use explicitly, which of the 3 approaches mentioned in #570 feels like the cleanest to me... but since I just posted a question there about whether I even understand how `use` is supposed to work in why3 correctly, you should probably check that you agree with me before merging this.

 I'm suspicious that I might be missing similar issues by doing

```rust
// Filter out enum variants, and maybe other things...
if !matches!(util::item_type(ctx.tcx, variant.def_id), ItemType::Unsupported(_))
```

 I haven't yet gone out of my way to try to find (or convince myself of the lack of) those similar issues. If you remove the check, a bunch of tests break with "unsupported definition kind" - but there might be a more principled version of the check available too.
 
 This PR is arguably a bit premature given the above, but I'm not going to have time to work on it for the next few days, so I thought I'd push it anyways.

xldenis commented on creusot/src/translation/function/place.rs:
```diff
@@ -180,11 +180,18 @@ pub fn translate_rplace_inner<'tcx>(
-                    ctx.translate_accessor(def.variants()[variant_id].fields[ix.as_usize()].did);
+                    let field_id = variant.fields[ix.as_usize()].did;
+                    // Filter out enum variants, and maybe other things...
+                    if !matches!(util::item_type(ctx.tcx, variant.def_id), ItemType::Unsupported(_))
+                    {
+                        names.insert(ctx.tcx.parent(field_id), subst);
+                    }
```
this can just be replaced by `names.insert(def.did(), subst)`, which will correctly add the edge for the datatype itself.
```suggestion
                    
                    names.insert(def.did(), subst);
                    
```

gmorenz commented:
Finally got back to this :)

Fixed the suggestion in the comment, also fixed another case of the same issue, and added a more complete test.

I haven't checked if #584 will completely obsolete this? Even if it does, it might be nice to get these expressions in the test suite so they don't break in the future.

xldenis commented on creusot/tests/should_succeed/bug/570.stderr:
```diff
@@ -0,0 +1,34 @@
+warning: struct `S2` is never constructed
+ --> 570.rs:8:8
+  |
+8 | struct S2 {
```
Can you fix these warnings by making things public?

gmorenz replied:
Done

xldenis reviewed (approved):
One last detail. 

As for #584, I wouldn't wait on  it. I'm going to land individual bits over the coming weeks, but I doubt #584 itself  will be merged
